servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -3192,7 +3192,67 @@ void updateRoleMemberUserAuthorityExpiry(final Role role, final String caller) {
         }
     }
 
-    void validateRoleMemberPrincipal(final String memberName, int principalType, final String userAuthorityFilter, final String caller) {
+    void validateUserPrincipal(final String memberName, boolean validateUserMember, final String userAuthorityFilter,
+                               final String caller) {
+
+        if (userAuthority == null) {
+            return;
+        }
+
+        if (validateUserMember) {
+            if (!userAuthority.isValidUser(memberName)) {
+                throw ZMSUtils.requestError("Principal " + memberName + " is not valid", caller);
+            }
+        }
+
+        // once we know it's a valid principal and we have a user
+        // authority filter configured, we'll check that as well
+        // if we're already determined that the principal is not
+        // valid there is no point of running this check
+
+        if (!StringUtil.isEmpty(userAuthorityFilter)) {
+            if (!ZMSUtils.isUserAuthorityFilterValid(userAuthority, userAuthorityFilter, memberName)) {
+                throw ZMSUtils.requestError("Invalid member: " + memberName +
+                        ". Required user authority filter not valid for the member", caller);
+            }
+        }
+    }
+
+    void validateServicePrincipal(final String memberName, final String caller) {
+
+        int idx = memberName.lastIndexOf('.');
+        if (idx == -1) {
+            throw ZMSUtils.requestError("Principal " + memberName + " is not valid", caller);
+        }
+
+        final String domainName = memberName.substring(0, idx);
+        final String serviceName = memberName.substring(idx + 1);
+
+        // first we need to check if the domain is on the list of
+        // our skip domains for service member validation. these
+        // are typically domains (like for ci/cd) where services
+        // are dynamic and do not need to be registered in Athenz
+
+        if (!validateServiceMemberSkipDomains.contains(domainName)) {
+            if (dbService.getServiceIdentity(domainName, serviceName, true) == null) {
+                throw ZMSUtils.requestError("Principal " + memberName + " is not a valid service", caller);
+            }
+        }
+    }
+
+    void validateGroupPrincipal(final String memberName, final String caller) {
+
+        int idx = memberName.indexOf(AuthorityConsts.GROUP_SEP);
+        final String domainName = memberName.substring(0, idx);
```

Comment:
we can have a check for idx == -1 here as well

Reply:
We don't need it. ZMSUtils.principalType sets the principal type to GROUP only if it contains the AuthorityConsts.GROUP_SEP value thus idx will never be -1. We do need it for service use case since the function treats non user/group types as service thus we need additional validation there.